error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/MockitoMockClassReferenceTest.java

```diff
@@ -0,0 +1,139 @@
+      CompilationTestHelper.newInstance(MockitoMockClassReference.class, getClass());
+  private final BugCheckerRefactoringTestHelper refactoringTestHelper =
+      BugCheckerRefactoringTestHelper.newInstance(MockitoMockClassReference.class, getClass());
+
+  @Test
+  void identification() {
+    compilationTestHelper
+        .addSourceLines(
+            "A.java",
+            "import static org.mockito.Mockito.mock;",
+            "import static org.mockito.Mockito.spy;",
+            "",
+            "import java.util.List;",
+            "",
+            "class A {",
+            "  // BUG: Diagnostic contains:",
+            "  String memberMock = mock(String.class);",
+            "  // BUG: Diagnostic contains:",
+            "  String memberSpy = mock(String.class);",
+            "",
+            "  void m() {",
+            "    Integer variableMock;",
+            "    // BUG: Diagnostic contains:",
+            "    variableMock = mock(Integer.class);",
+            "    // BUG: Diagnostic contains:",
+            "    List nonGenericallyTypedMock = mock(List.class);",
+            "    var dynamicallyTypedMock = mock(Integer.class);",
+            "    Integer namedMock = mock(Integer.class, \"name\");",
+            "    Object subtypeMock = mock(Integer.class);",
+            "",
+            "    // BUG: Diagnostic contains:",
+            "    String equalTypedSpy = spy(String.class);",
+            "    // BUG: Diagnostic contains:",
+            "    List nonGenericallyTypedSpy = spy(List.class);",
+            "    var dynamicallyTypedSpy = spy(Integer.class);",
+            "    Object objectSpy = spy(new Object());",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  void unimplementedCases() {
+    // XXX: Move to identification once fixed.
+    compilationTestHelper
+        .addSourceLines(
+            "A.java",
+            "import static org.mockito.Mockito.mock;",
+            "",
+            "import java.math.BigInteger;",
+            "import java.util.List;",
+            "",
+            "class A {",
+            "  void m() {",
+            "    // This case is arguable as it is unsafe in any case.",
+            "    // This currently is not identified as the `erasure` is the same.",
+            "    List<String> genericallyTypedMock = mock(List.class);",
+            "",
+            "    // This case is problematic and should not be replaced.",
+            "    // The real type of `variableTypedMock` is `BigInteger`.",
+            "    // But when replaced, it will be `Number`.",
+            "    Class<? extends Number> variableType = BigInteger.class;",
+            "    Number variableTypedMock = mock(variableType);",
+            "  }",
+            "",
+            "  <T extends Number> T getGenericMock(Class<T> clazz) {",
+            "    // No idea how to detect type T as part of generics usage.",
+            "    return mock(clazz);",
+            "  }",
```

Comment:
Coming to think of it, if it were not such a complex case where `T extends Number`, but where it was just `T`, then we could likely rewrite it:

```diff
  <T> T getGenericMock(Class<T> clazz) {
-    return mock(clazz);
+    return mock();
  }
```

Perhaps this case is just similar as `variableTypedMock` above but instead with a generic result type. So we might be able to simplify these test cases a bit and indeed follow the `ASTHelpers.isSameType` approach where it compared the `erasure` of both types.